src/ts/form_render/formRender.ts

```diff
@@ -0,0 +1,365 @@
+        if (!formData) {
+          formData = this.options.formData
+        }
+
+        this.options = Object.assign({}, this.options, options, {
+          formData: this.parseFormData(formData),
+        })
+
+        this.render()
+      },
+      html: () => $(this.el).html(),
+    }
+  }
+
+  private registerCustomTemplates() {
+    if (Object.keys(this.options.templates).length) {
+      controlCustom.registerCustom(this.options.templates)
+    }
+  }
+
+  /**
+   * Clean up passed object configuration to prepare for use with the markup function
+   * @param {Object} field - object of field configuration
+   * @param {Number} instanceIndex - instance index
+   * @return {Object} sanitized field object
+   */
+  santizeField(field: Field, instanceIndex?: number): Field {
+    const sanitizedField = Object.assign({}, field)
+    if (instanceIndex) {
+      sanitizedField.id = field.id && `${field.id}-${instanceIndex}`
+      sanitizedField.name = field.name && `${field.name}-${instanceIndex}`
+    }
+
+    sanitizedField.className = Array.isArray(field.className)
+      ? unique(field.className.join(' ').split(' ')).join(' ')
+      : field.className || field.class || null
+
+    delete sanitizedField.class
+
+    if (field.values) {
+      field.values = (field.values as []).map(option => trimObj(option))
+    }
+
+    return trimObj(sanitizedField)
+  }
+
+  /**
+   * parses `container` option or returns element
+   * @param  {Object} element
+   * @return {Object} parsedElement
+   */
+  getElement(element) {
+    element = this.options.container || element
+    if (element instanceof jQuery) {
+      element = element[0]
+    } else if (typeof element === 'string') {
+      element = document.querySelector(element)
+    }
+    return element
+  }
+
+  /**
+   * Main render method which produces the form from passed configuration
+   * @param {Object} element - an html element to render the form into (optional)
+   * @param {Number} instanceIndex - instance index
+   * @return {Object} rendered form
+   */
+  render(instanceIndex = 0) {
+    const opts = this.options
+    this.el = this.getElement(this.el)
+
+    if (opts.formData) {
+      const rendered: HTMLElement[] = []
+
+      const engine = new Layout(opts.layoutTemplates)
+
+      for (let i = 0; i < opts.formData.length; i++) {
+        const fieldData: Field = opts.formData[i]
+        const sanitizedField = this.santizeField(fieldData, instanceIndex)
+
+        const controlClass = control.getRegisteredClassControl(fieldData.type, fieldData.subtype)
+        const field = engine.build(controlClass, sanitizedField)
+
+        rendered.push(field)
+      }
+
+      //Kevin -- what is instanceIndex all about? If you had multiple forms wouldnt you load this separately?
+      if (this.el) {
+        this.instanceContainers[instanceIndex] = this.el
+      }
+
+      //Kevin -- why would we not be rendering? does this really need an if/else?
```

Comment:
sometimes an element is not available to render to and there was a feature request to be able to affect an instance prior to rendering it. sometimes they just want the html